store.rb

```diff
@@ -118,4 +153,26 @@ def get_node(node)
       raise e
     end
   end
+
+  # immediately update cache if node joins/leaves
+  def watch()
+    return if @zk.nil?
+    @root_watcher ||= @zk.register("/", :only => :child) do |event|
+      @log.info "Children added/deleted"
+      reload_instances
+    end
+  end
+
+  def reload_instances()
+    instances = load_instances_from_zk()
+    cache_instances(instances)
+  end
+
+  def get_cached_instances()
+    @mutex.synchronize { @cached_nodes }
+  end
+
+  def cache_instances(instances)
+    @mutex.synchronize { @cached_nodes = instances }
```

Comment:
Makes sense! With MRI for a single statement it's atomic. As for wrapping zk fetch and assignment together with mutex. I think we can use a monotonically increasing sequence as a work around.